adr/cdk-cli-lib.md

```diff
@@ -0,0 +1,414 @@
+  async produce(context: Record<string, any>) {
+    const app = new cdk.App({ context });
+    const stack = new cdk.Stack(app);
+    return app.synth().directory;
+  }
+}
+
+const cli = AwsCdkCli.fromCloudAssemblyDirectoryProducer(new MyProducer());
+await cli.synth();
+```
+
+This library requires us to create the `cdk.App` within the `produce` method of
+`ICloudAssemblyDirectoryProducer`. This is because the `App` and all constructs
+within it must be constructed with the full `context` value. It is not possible
+to add context after a construct has been constructed.
+
+2. Because the constructs can be called multiple times, any Pulumi resources
+   which are created inside a construct class will also be constructed multiple
+   times. This means we need a way of knowing when the final call happens and
+   it is safe to construct the Pulumi resources.
+   1. One thing to note: CDK lookups are infrequently performed. Typically they
+      are performed once when they are added and then they are cached in the
+      `cdk.context.json` file. This means that the multiple pass execution will
+      happen infrequently.
+
+
+## Decision
+
+### App with callback argument
+
+We will introduce a new `App` class which accepts a callback function as the
+second argument. The user would have to create resources within this call back
+function. The API that looks something like this for the end user. 
+
+This would allow us to create the `App` inside a Pulumi `ComponentResource` and
+then pass the created `App` to the `(scope: pulumicdk.App) => {}` function.
+
+```javascript
+const app = new pulumicdk.App('app', (scope: pulumicdk.App) => {
+    new Ec2CdkStack(scope, 'teststack');
+});
+
+// with app outputs
+const app = new pulumicdk.App('app', (scope: pulumicdk.App): pulumicdk.AppOutputs => {
+    const stack = new Ec2CdkStack(scope, 'teststack');
+    return {
+        instanceId: stack.instanceId,
+    };
+});
+export const instanceId = app.outputs['instanceId'];
+```
+
+### User retry
+
+In order to handle #2 above where resources will be called multiple times, we
+will require the user to run the Pulumi command two times. Context
+lookups are different from Pulumi data calls since they are meant to be called
+one time when they are initially added and then from then on the data is cached
+in `cdk.context.json` and the lookup doesn’t need to be performed again. The
+user experience would look something like this:
+
+1. User adds a lookup `Vpc.fromLookup()`
+2. `pulumi up`
+3. Lookups are performed, stored in `cdk.context.json` and an error is thrown.
+4. `pulumi up` succeeds from then on.
+
+While this option is not ideal, it will offer the best compromise which still
+allows users to perform all CDK lookups. We will call it out in documentation
+and recommend users to use Pulumi native lookups for most things and only fall
+back to CDK lookups for `VPC` and `Load Balancers`.
+
+## Alternatives
+
+### Mock resource calls
```

Comment:
Should we cut an issue for automatically handling context lookups? I think that would be great to gather user inputs and see if it's actually necessary

Reply:
https://github.com/pulumi/pulumi-cdk/issues/184

Reply:
Ah nice!